Add support for os_signpost for WebContent on iOS

bnham opened the pull request:
#### 2242351e265d11469993410984fdda3ec21e6de2
<pre>
Add support for os_signpost for WebContent on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=264357">https://bugs.webkit.org/show_bug.cgi?id=264357</a>
<a href="https://rdar.apple.com/117593169">rdar://117593169</a>

Reviewed by Per Arne Vollan.

269026@main blocked WebContent from talking to logd. This also broke os_signpost in WebContent since
signposts are emitted to logd.

To work around this, in this patch we send WebContent os_signposts down the same IPC pipe as regular
os_logs. On the receiving side (in NetworkProcess), we deserialize the log line and emit an
os_signpost. This requires some unfortunate use of scanf on the receiving side, but I don&apos;t see a
better way of doing this without improved signpost hooks from libtrace (<a href="https://rdar.apple.com/111422207).">rdar://111422207).</a>

This also requires a bunch of C macros because signpost names *must* be C string literal tokens from
the point of view of the C preprocessor, due to the way os_signpost is implemented. So there are
several places here where we use a switch-case through every single possible signpost name in order
to emit the right os_signpost call. The various os_signpost APIs themselves are also macros, so
there are also places where we have to switch-case through every signpost API type (emit event,
begin interval, and end inverval) in order to call the appropriate os_signpost macro.

I also made some changes to few existing callers of os_signpost in the codebase:

- There were two places where we were calling the EmitSignpostAlways variant of the API. This
  probably doesn&apos;t make sense anymore considering a signpost is now heavier weight and requires an
  IPC. I converted those call sites to use `os_signpost` directly, which means they emit signposts
  when logd is enabled (on Mac), and don&apos;t emit signposts when logd is blocked.

- `handleWheelEvent` used the animation tagged variant of the signpost API. This is hard to continue
  using when sending signposts over IPC since it requires the format string to be encoded in a
  particular manner. I also converted this to use os_signpost directly, which should be okay since
  this code is only running on Mac.

- I consolidated all of the page loading and paint signposts under a single signpost name called
  NavigationAndPaintTiming. For each document loaded, we now emit an interval from [startTime,
  loadEventEnd) under this signpost name. Within this interval, key milestones from the
  NavigationTiming and PaintTiming APIs are annotated, e.g. domInteractive,
  domContentLoadedEventStart, loadEventStart, firstContentfulPaint, etc.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(beginSignpostInterval):
(endSignpostInterval):
(emitSignpostEvent):
(emitSignpost):
(WTFSignpostHandleIndirectLog):
(machTimebaseInfo):
(WTFCurrentContinuousTime):
* Source/WebCore/dom/Document.cpp:
(WebCore::m_frameIdentifier):
(WebCore::Document::~Document):
(WebCore::Document::setReadyState):
(WebCore::Document::enqueuePaintTimingEntryIfNeeded):
(WebCore::Document::finishedParsing):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::dispatchLoadEvent):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::firePaintRelatedMilestonesIfNeeded):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didSendData):
(WebKit::NetworkDataTaskCocoa::didReceiveChallenge):
(WebKit::NetworkDataTaskCocoa::didCompleteWithError):
(WebKit::NetworkDataTaskCocoa::didReceiveData):
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::cancel):
(WebKit::NetworkDataTaskCocoa::resume):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: <a href="https://commits.webkit.org/270471@main">https://commits.webkit.org/270471@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/709906dc74c89f5adf781f24effcdfbca2884573

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25312 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3853 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26567 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27422 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23216 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25584 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5572 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23417 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25554 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21838 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28001 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2533 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22779 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28882 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22012 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23083 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26720 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/24541 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2491 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/774 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31973 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3848 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6989 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6137 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2933 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2825 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
```diff
@@ -1482,9 +1486,17 @@ void NetworkConnectionToWebProcess::logOnBehalfOfWebContent(IPC::DataReference&&
+    auto osLogPointer = osLog.get() ? osLog.get() : OS_LOG_DEFAULT;
+    auto logData = reinterpret_cast<const char*>(logString.data());
+
+#if HAVE(OS_SIGNPOST)
+    if (WTFSignpostHandleIndirectLog(osLogPointer, pid, logData))
+        return;
+#endif
```
Would it make sense to only enable this code in internal builds, in order to avoid the sscanf for non internal builds?

bnham replied:
I think we need to be able to get signposts on internal builds that are emulating customer builds, so we don't want to do an internal build check here.

`WTFSignpostHandleIndirectLog` doesn't do anything unless the log subsystem is `com.apple.WebKit` and the category is `Signposts` so I expect that it will do nothing nearly all of the time.

pvollan reviewed (approved):
R=me.

pvollan commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -195,72 +200,109 @@ using WTF::tracePoint;
+        emitMacro(wtfHandle, wtfSignpostID, #name, format, ##__VA_ARGS__); \
+    } while (0)
+
+#define WTFEmitSignpostIndirectlyWithType(type, pointer, name, timeDelta, format, ...) \
+    os_log(WTFSignpostLogHandle(), "type=%d name=%d p=%" PRIuPTR " ts=%llu " format, type, WTFOSSignpostName ## name, reinterpret_cast<uintptr_t>(pointer), WTFCurrentContinuousTime(timeDelta), ##__VA_ARGS__)
```
Another option here could perhaps be to delegate this to a signpost observer, that would send a signpost message to the Networking process. Then you could avoid the sscanf in the Networking process.

bnham replied:
I don't actually see a way to do this without invoking `os_log`. That's because `os_signpost` takes an `os_log`-compatible format string and I don't see any API that can actually expand that format string outside of calling `os_log` and hooking into the object passed to `os_log_hook`.

bnham committed:
Add support for os_signpost for WebContent on iOS
https://bugs.webkit.org/show_bug.cgi?id=264357
rdar://117593169

Reviewed by Per Arne Vollan.

269026@main blocked WebContent from talking to logd. This also broke os_signpost in WebContent since
signposts are emitted to logd.

To work around this, in this patch we send WebContent os_signposts down the same IPC pipe as regular
os_logs. On the receiving side (in NetworkProcess), we deserialize the log line and emit an
os_signpost. This requires some unfortunate use of scanf on the receiving side, but I don't see a
better way of doing this without improved signpost hooks from libtrace (rdar://111422207).

This also requires a bunch of C macros because signpost names *must* be C string literal tokens from
the point of view of the C preprocessor, due to the way os_signpost is implemented. So there are
several places here where we use a switch-case through every single possible signpost name in order
to emit the right os_signpost call. The various os_signpost APIs themselves are also macros, so
there are also places where we have to switch-case through every signpost API type (emit event,
begin interval, and end inverval) in order to call the appropriate os_signpost macro.

I also made some changes to few existing callers of os_signpost in the codebase:

- There were two places where we were calling the EmitSignpostAlways variant of the API. This
  probably doesn't make sense anymore considering a signpost is now heavier weight and requires an
  IPC. I converted those call sites to use `os_signpost` directly, which means they emit signposts
  when logd is enabled (on Mac), and don't emit signposts when logd is blocked.

- `handleWheelEvent` used the animation tagged variant of the signpost API. This is hard to continue
  using when sending signposts over IPC since it requires the format string to be encoded in a
  particular manner. I also converted this to use os_signpost directly, which should be okay since
  this code is only running on Mac.

- I consolidated all of the page loading and paint signposts under a single signpost name called
  NavigationAndPaintTiming. For each document loaded, we now emit an interval from [startTime,
  loadEventEnd) under this signpost name. Within this interval, key milestones from the
  NavigationTiming and PaintTiming APIs are annotated, e.g. domInteractive,
  domContentLoadedEventStart, loadEventStart, firstContentfulPaint, etc.

* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/cocoa/SystemTracingCocoa.cpp:
(beginSignpostInterval):
(endSignpostInterval):
(emitSignpostEvent):
(emitSignpost):
(WTFSignpostHandleIndirectLog):
(machTimebaseInfo):
(WTFCurrentContinuousTime):
* Source/WebCore/dom/Document.cpp:
(WebCore::m_frameIdentifier):
(WebCore::Document::~Document):
(WebCore::Document::setReadyState):
(WebCore::Document::enqueuePaintTimingEntryIfNeeded):
(WebCore::Document::finishedParsing):
* Source/WebCore/dom/ScriptElement.cpp:
(WebCore::ScriptElement::executeClassicScript):
(WebCore::ScriptElement::registerImportMap):
(WebCore::ScriptElement::executeModuleScript):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::dispatchLoadEvent):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::firePaintRelatedMilestonesIfNeeded):
* Source/WebCore/platform/mac/ScrollingEffectsController.mm:
(WebCore::ScrollingEffectsController::handleWheelEvent):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::logOnBehalfOfWebContent):
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::~NetworkDataTaskCocoa):
(WebKit::NetworkDataTaskCocoa::didSendData):
(WebKit::NetworkDataTaskCocoa::didReceiveChallenge):
(WebKit::NetworkDataTaskCocoa::didCompleteWithError):
(WebKit::NetworkDataTaskCocoa::didReceiveData):
(WebKit::NetworkDataTaskCocoa::didReceiveResponse):
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
(WebKit::NetworkDataTaskCocoa::cancel):
(WebKit::NetworkDataTaskCocoa::resume):
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::registerLogHook):

Canonical link: https://commits.webkit.org/270471@main
